flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonFileSystemFormatFactory.java

```diff
@@ -0,0 +1,263 @@
+	 * A {@link JsonInputFormat} is responsible to read {@link RowData} records
+	 * from json format files.
+	 */
+	public static class JsonInputFormat extends DelimitedInputFormat<RowData> {
+		/**
+		 * Code of \r, used to remove \r from a line when the line ends with \r\n.
+		 */
+		private static final byte CARRIAGE_RETURN = (byte) '\r';
+
+		/**
+		 * Code of \n, used to identify if \n is used as delimiter.
+		 */
+		private static final byte NEW_LINE = (byte) '\n';
+
+		private final List<DataType> fieldTypes;
+		private final List<String> fieldNames;
+		private final int[] selectFields;
+		private final List<String> partitionKeys;
+		private final String defaultPartValue;
+		private final long limit;
+		private final List<String> jsonSelectFieldNames;
+		private final int[] jsonFieldMapping;
+		private final JsonRowDataDeserializationSchema deserializationSchema;
+
+		private transient boolean end;
+		private transient long emitted;
+		// reuse object for per record
+		private transient GenericRowData rowData;
+
+		public JsonInputFormat(
+				Path[] filePaths,
+				List<DataType> fieldTypes,
+				List<String> fieldNames,
+				int[] selectFields,
+				List<String> partitionKeys,
+				String defaultPartValue,
+				long limit,
+				List<String> jsonSelectFieldNames,
+				int[] jsonFieldMapping,
+				JsonRowDataDeserializationSchema deserializationSchema) {
+			super.setFilePaths(filePaths);
+			this.fieldTypes = fieldTypes;
+			this.fieldNames = fieldNames;
+			this.selectFields = selectFields;
+			this.partitionKeys = partitionKeys;
+			this.defaultPartValue = defaultPartValue;
+			this.limit = limit;
+			this.jsonSelectFieldNames = jsonSelectFieldNames;
+			this.jsonFieldMapping = jsonFieldMapping;
+			this.deserializationSchema = deserializationSchema;
+		}
+
+		@Override
+		public boolean supportsMultiPaths() {
+			return true;
+		}
+
+		@Override
+		public void open(FileInputSplit split) throws IOException {
+			super.open(split);
+			this.end = false;
+			this.emitted = 0L;
+			this.rowData = PartitionPathUtils.fillPartitionValueForRecord(fieldNames, fieldTypes, selectFields,
+				partitionKeys, currentSplit.getPath(), defaultPartValue);
+		}
+
+		@Override
+		public boolean reachedEnd() {
+			return emitted >= limit || end;
+		}
+
+		@Override
+		public RowData readRecord(RowData reuse, byte[] bytes, int offset, int numBytes) throws IOException {
+			GenericRowData returnRecord = rowData;
+
+			if (this.getDelimiter() != null && this.getDelimiter().length == 1
+				&& this.getDelimiter()[0] == NEW_LINE && offset + numBytes >= 1
+				&& bytes[offset + numBytes - 1] == CARRIAGE_RETURN){
+				numBytes -= 1;
+			}
+			byte[] trimBytes = Arrays.copyOfRange(bytes, offset, offset + numBytes);
+			GenericRowData jsonRow = (GenericRowData) deserializationSchema.deserialize(trimBytes);
+			if (jsonRow != null) {
+				for (int i = 0; i < jsonSelectFieldNames.size(); i++) {
+					returnRecord.setField(jsonFieldMapping[i], jsonRow.getField(i));
```

Comment:
`jsonRow.getField`? `jsonSelectFieldNames` useless?
You can add tests too.

Reply:
jsonSelectFieldNames length may less than jsonRow.size()?

Reply:
Do you want the project fields? But only the fields in front?

Reply:
got your point，use:
```
for (int i = 0; i < jsonFieldMapping.length; i++) {
	returnRecord.setField(jsonFieldMapping[i], jsonRow.getField(i));
}
```

Reply:
It looks like totally wrong, you could add some projection tests.